File: dashboard/session_sync.py
"""
The one small file the dashboard and the Unity viewer both watch (item N1).

WHY A FILE AND NOT A SOCKET. Neither process may be able to wedge the other.
A socket makes one of them a server and the other a client, which means a
startup order, a port to collide, a connection to drop and a reconnect loop to
get wrong; and the moment the viewer blocks on a read, a dashboard that is busy
stepping a year has stalled the editor. A file has none of that. Both sides
poll it, both sides tolerate it being absent, and if one of them is closed the
other simply sees a value that stops changing.

WHY NOT INSIDE THE BUNDLE, which is the obvious place. The bundle Unity reads
is a COPY: `outputs/unity/village` is installed into the consuming project's
`Assets/StreamingAssets/extnpc/<world>/`, so a file written into one is not the
file read from the other. That copy is easy to forget and produces the worst
kind of bug, where both sides are working perfectly on different files. So the
sync file lives at a fixed user-level path that neither side owns, outside both
trees, overridable with `EXTNPC_SESSION_FILE` for tests and for a second
concurrent session.

WHAT IT CARRIES. A selection and a year, which FINAL_LINE N1 says are one
mechanism and are done together: they are both "what am I looking at", they
change together when you click a villager in a past year, and splitting them
into two files would let them disagree.

WHAT IT DOES NOT CARRY, and this is N2's decision rather than an omission. It
does not step the simulation. Engine import is 12.4 s and a tick is ~111 ms at
67 living, so roughly a second per simulated year at 600 people; export then
view is the right shape for this workload. This file points two viewers at the
same thing. It does not make Unity a client of the engine.

THE ECHO PROBLEM, which is the one real subtlety. Both sides write and both
sides read, so each will read back its own writes and, if it applies them,
fight the user's input. Every payload therefore names its `source`, and a
reader ignores anything it wrote itself. `seq` increases so a reader can also
tell a genuinely new message from the same one polled twice, which matters
because two writes inside the same clock tick are otherwise indistinguishable.
"""

from __future__ import annotations

import json
import os
import tempfile
import time
from typing import Any, Dict, Optional

__all__ = [
    "SCHEMA",
    "SessionState",
    "default_path",
    "read_state",
    "write_state",
]

SCHEMA = 1

#: Sentinel for "this client has published nothing yet". A module-level object
#: rather than None, because (None, None) is a legitimate state to publish.
_NOTHING_PUBLISHED = object()

#: Who wrote a payload. A reader ignores its own source.
SOURCE_DASHBOARD = "dashboard"
SOURCE_UNITY = "unity"


def default_path() -> str:
    """Where the sync file lives when nobody says otherwise.

    A user-level path rather than a temp directory: a session-scoped temp
    directory is cleaned up underneath a long-running editor, and the two
    processes start at different times, so "the temp dir" is not necessarily
    the same directory for both of them.
    """
    override = os.environ.get("EXTNPC_SESSION_FILE")
    if override:
        return override
    return os.path.join(os.path.expanduser("~"), ".extnpc", "session.json")


class SessionState:
    """What both viewers are looking at.

    `year` is the simulated year being VIEWED, which is not always the newest
    one: the dashboard calls that time travel and stores None for "live". None
    is preserved rather than collapsed to the latest tick, because "follow the
    front" and "hold at year 40" are different intentions and the viewer on the
    other side should be able to honour the difference.
    """

    __slots__ = ("selected", "year", "world", "source", "seq", "written_ms")

    def __init__(self, selected: Optional[str] = None,
                 year: Optional[int] = None,
                 world: str = "",
                 source: str = "",
                 seq: int = 0,
                 written_ms: int = 0) -> None:
        self.selected = selected
        self.year = year
        self.world = world
        self.source = source
        self.seq = int(seq)
        self.written_ms = int(written_ms)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "schema": SCHEMA,
            "selected": self.selected,
            "year": self.year,
            "world": self.world,
            "source": self.source,
            "seq": self.seq,
            "written_ms": self.written_ms,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "SessionState":
        return cls(
            selected=data.get("selected"),
            year=data.get("year"),
            world=data.get("world") or "",
            source=data.get("source") or "",
            seq=data.get("seq") or 0,
            written_ms=data.get("written_ms") or 0,
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, SessionState):
            return NotImplemented
        return (self.selected == other.selected and self.year == other.year
                and self.world == other.world)

    def __repr__(self) -> str:
        return (f"SessionState(selected={self.selected!r}, year={self.year!r}, "
                f"world={self.world!r}, source={self.source!r}, seq={self.seq})")


def read_state(path: Optional[str] = None) -> Optional[SessionState]:
    """The current state, or None when there is nothing usable to read.

    NEVER RAISES, and that is the contract rather than defensive habit. This is
    called from a Dash callback on a timer. An exception here would take out
    the callback that also advances the world, so a missing file, a half-written
    one, a file being replaced under us or a payload from a future schema all
    have to mean the same thing: no news. The alternative is a viewer that dies
    because the other viewer was mid-write.
    """
    target = path or default_path()
    try:
        with open(target, encoding="utf-8") as fh:
            data = json.load(fh)
    except (OSError, ValueError):
        return None

    if not isinstance(data, dict):
        return None
    # An unknown schema is ignored rather than guessed at. Reading a future
    # field layout as if it were this one is how two versions of the same tool
    # silently disagree about which villager is selected.
    if data.get("schema") != SCHEMA:
        return None
    return SessionState.from_dict(data)


def write_state(state: SessionState, path: Optional[str] = None) -> bool:
    """Write atomically. Returns False rather than raising when it cannot.

    ATOMIC BECAUSE THE READER IS POLLING. A plain open-and-write leaves a
    window in which the file exists and is half a JSON document, and the other
    side is reading on a timer, so it WILL land in that window eventually. The
    temp-file-then-replace means the target only ever holds a payload that was
    written completely: an interrupted write damages the temp file, which is
    then removed, and the previous payload survives untouched.

    WHAT IT DOES NOT BUY, measured on Windows on 2026-08-24 rather than
    assumed. It does not make writes invisible to a concurrent reader. Under a
    reader polling flat out, failed reads went from about 35% of attempts with
    a plain write to about 9% with this one -- a real improvement, and not
    zero, because `os.replace` can collide with a reader that currently has the
    file open and make its `open()` fail. That surfaces as OSError, which
    `read_state` already reports as "no news", and no news is harmless to a
    poller: the next tick reads the new payload. The claim here is that the
    file is never CORRUPT, not that a read never fails.

    `os.replace` rather than `os.rename`: on Windows, rename onto an existing
    file raises, which would make every write after the first one fail.
    """
    target = path or default_path()
    directory = os.path.dirname(target)
    try:
        if directory:
            os.makedirs(directory, exist_ok=True)
        state.written_ms = int(time.time() * 1000)

        handle, temp = tempfile.mkstemp(
            dir=directory or None, prefix=".session-", suffix=".tmp")
        try:
            with os.fdopen(handle, "w", encoding="utf-8", newline="\n") as fh:
                json.dump(state.to_dict(), fh, indent=1)
                fh.write("\n")
            os.replace(temp, target)
        except BaseException:
            # Do not leave a temp file behind on a failed write; the directory
            # is long-lived and this runs on a timer.
            try:
                os.unlink(temp)
            except OSError:
                pass
            raise
        return True
    except OSError:
        return False


class SyncClient:
    """One side of the link: writes its own state, ignores its own echo.

    Holds the sequence number and the last state it published, so a caller can
    poll on a timer and be told only about genuine changes made by the OTHER
    side. Keeping that bookkeeping here rather than in the Dash callback is
    what stops the callback from becoming a feedback loop, which is the same
    failure `slider-echo` already exists to prevent in `app.py`.
    """

    def __init__(self, source: str, world: str = "",
                 path: Optional[str] = None) -> None:
        self.source = source
        self.world = world
        self.path = path or default_path()
        self._seq = 0
        # (source, seq) of the last payload this client accepted as news.
        #
        # THE PAIR, NOT THE SEQ. `seq` counts each writer's own messages, so
        # the two counters collide: the dashboard's third message and the
        # viewer's third message are both seq 3. Comparing a bare seq made a
        # genuine message from the other side look like one already seen, and
        # the sync silently stopped working as soon as BOTH sides had spoken.
        # Every unit test passed, because each one had only one side talking.
        self._last_seen = None
        # The last (selected, year) this side put on the wire, or a sentinel
        # meaning "nothing yet". A caller polling on a timer uses it to avoid
        # rewriting an unchanged state two or three times a second forever; the
        # sentinel is a distinct object rather than None so that publishing
        # (None, None) -- nobody selected, following the newest year, which is
        # the dashboard's opening state -- still counts as having published.
        self._last_published = _NOTHING_PUBLISHED

    @property
    def last_published(self):
        """What this side last put on the wire, as `(selected, year)`.

        Compares unequal to every real pair until something has been published,
        so the first call always writes.
        """
        return self._last_published

    def publish(self, selected: Optional[str], year: Optional[int]) -> bool:
        """Announce what this side is looking at."""
        self._seq += 1
        state = SessionState(selected=selected, year=year, world=self.world,
                             source=self.source, seq=self._seq)
        ok = write_state(state, self.path)
        if ok:
            self._last_published = (selected, year)
        return ok

    def poll(self) -> Optional[SessionState]:
        """What the OTHER side is looking at, or None when there is no news.

        Returns None for: no file, unreadable file, our own writes, a payload
        already returned, and a payload for a different world.
        """
        state = read_state(self.path)
        if state is None:
            return None
        # Our own writes, including those of a previous process: after a
        # restart `seq` is back at zero while the file still holds the old
        # payload, and only this check stops it being read back as news.
        if state.source == self.source:
            return None
        if (state.source, state.seq) == self._last_seen:
            return None
        # A dashboard running a different world must not drag the viewer to a
        # villager who does not exist in the one on screen. Empty means "not
        # stated", which is allowed so a caller can opt out.
        if self.world and state.world and state.world != self.world:
            return None
        self._last_seen = (state.source, state.seq)
        # Record what we heard as though we had published it. Without this the
        # caller applies the incoming state, sees on its next tick that its own
        # state differs from `last_published`, and sends it straight back. The
        # other side drops that as a duplicate rather than looping forever, so
        # the symptom is not a hang: it is two processes writing the same file
        # to each other several times a second for no reason. The C# bridge
        # does the same thing after applying a payload.
        self._last_published = (state.selected, state.year)
        return state

File: unity/com.samal.extnpc/Runtime/Sync/SessionSync.cs
using System;
using System.Globalization;
using System.IO;
using Newtonsoft.Json.Linq;
using UnityEngine;

namespace ExtNPC.Sync
{
    /// <summary>
    /// One side of the dashboard link: what both viewers are looking at (N1).
    ///
    /// The C# half of `dashboard/session_sync.py`. Read that file for the
    /// reasoning; this one only has to agree with it. The agreement is a JSON
    /// object with `schema`, `selected`, `year`, `world`, `source` and `seq`,
    /// and `tests/test_session_sync.py` plus <c>SessionSyncTests</c> pin the
    /// same shape from both ends.
    ///
    /// WHY A FILE. Neither process may be able to wedge the other. A socket
    /// would make one of them a server, with a startup order, a port to
    /// collide and a reconnect loop; and an editor blocking on a read from a
    /// dashboard that is busy stepping a year is a frozen editor. Both sides
    /// poll a file, both tolerate its absence, and a closed dashboard looks
    /// like a value that stopped changing.
    ///
    /// IT DOES NOT STEP THE SIMULATION. FINAL_LINE N2 measured that: engine
    /// import is 12.4 s and a tick runs ~111 ms at 67 living, so roughly a
    /// second per simulated year at 600 people. Export then view is the right
    /// shape for this workload. This points two viewers at the same villager
    /// and the same year; it does not make Unity a client of the engine.
    ///
    /// THE ECHO PROBLEM. Both sides write and both sides read, so each will
    /// read back its own writes and, if it applies them, fight the user. Every
    /// payload names its `source` and a reader drops its own. `seq` then tells
    /// a new message from the same one polled twice. Both are needed: within a
    /// process the seq is enough, but after a RESTART the seq resets to zero
    /// while the file still holds the old process's payload, and only the
    /// source check stops that being read back as news.
    /// </summary>
    public sealed class SessionSync
    {
        public const int Schema = 1;
        public const string SourceUnity = "unity";
        public const string SourceDashboard = "dashboard";

        /// <summary>What the other side is looking at.</summary>
        public struct State
        {
            /// <summary>Selected villager, or null for no selection. Null is a
            /// MESSAGE, not a silence: deselecting has to travel or the other
            /// side keeps the old villager highlighted forever.</summary>
            public string Selected;

            /// <summary>The year being viewed, or null for "follow the newest".
            /// Not collapsed to the latest tick, because "follow the front" and
            /// "hold at year 40" are different intentions.</summary>
            public int? Year;

            public string World;
            public string Source;
            public long Seq;
        }

        private readonly string _source;
        private readonly string _world;
        private long _seq;

        // (source, seq) of the last payload accepted as news.
        //
        // THE PAIR, NOT THE SEQ. `seq` counts each writer's OWN messages, so
        // the two counters collide: the dashboard's first message and this
        // viewer's first message are both seq 1. Comparing a bare seq made a
        // genuine message look like one already seen, and the link went deaf
        // the moment BOTH sides had spoken. Found end to end on 2026-08-24;
        // every unit test on both sides passed, because each had only one side
        // talking.
        private string _lastSeenSource;
        private long _lastSeenSeq = -1;

        public string Path { get; private set; }

        public SessionSync(string source, string world, string path = null)
        {
            _source = source;
            _world = world ?? "";
            Path = string.IsNullOrEmpty(path) ? DefaultPath() : path;
        }

        /// <summary>
        /// Where the file lives when nobody says otherwise.
        ///
        /// NOT INSIDE THE BUNDLE, which is the obvious place and is wrong. The
        /// bundle here is a COPY: the exporter writes `outputs/unity/&lt;world&gt;`
        /// and that is installed into `Assets/StreamingAssets/extnpc/&lt;world&gt;`,
        /// so a file written into one is not the file read from the other. Both
        /// sides would work perfectly on different files, which is the worst
        /// kind of bug to be looking at. A user-level path outside both trees
        /// has no such copy.
        /// </summary>
        public static string DefaultPath()
        {
            string over = Environment.GetEnvironmentVariable("EXTNPC_SESSION_FILE");
            if (!string.IsNullOrEmpty(over)) return over;

            string home = Environment.GetFolderPath(
                Environment.SpecialFolder.UserProfile);
            if (string.IsNullOrEmpty(home))
                home = Environment.GetEnvironmentVariable("HOME") ?? ".";
            return System.IO.Path.Combine(home, ".extnpc", "session.json");
        }

        /// <summary>
        /// Announce what this side is looking at. False when it could not be
        /// written, which is not worth an exception: the next change tries
        /// again and the other side simply has stale news in the meantime.
        /// </summary>
        public bool Publish(string selected, int? year)
        {
            _seq++;
            var payload = new JObject
            {
                ["schema"] = Schema,
                ["selected"] = selected == null ? JValue.CreateNull() : new JValue(selected),
                ["year"] = year.HasValue ? new JValue(year.Value) : JValue.CreateNull(),
                ["world"] = _world,
                ["source"] = _source,
                ["seq"] = _seq,
                ["written_ms"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            };

            try
            {
                string dir = System.IO.Path.GetDirectoryName(Path);
                if (!string.IsNullOrEmpty(dir)) Directory.CreateDirectory(dir);

                // Same temp-then-move as the Python side, for the same reason:
                // the other side is polling, so an interrupted write must
                // damage a temp file rather than the payload anyone is reading.
                //
                // NOT A GUID, and the reason is a project rule rather than a
                // preference. `test_the_visual_layer_draws_no_random_numbers`
                // forbids `Guid.NewGuid` anywhere under Runtime, because
                // invariant 5 forbids the viewer inventing variance and the
                // rule is deliberately absolute rather than case-by-case. A
                // temp filename is not visual variance, but arguing the
                // exception would cost the rule its teeth, and nothing here
                // actually needs randomness: the source names the writer and
                // the sequence number is already unique per message, so the two
                // together are unique without drawing anything.
                string temp = Path + ".tmp-" + _source + "-" + _seq;
                try
                {
                    File.WriteAllText(temp, payload.ToString());
                    if (File.Exists(Path)) File.Delete(Path);
                    File.Move(temp, Path);
                }
                catch
                {
                    try { if (File.Exists(temp)) File.Delete(temp); } catch { }
                    throw;
                }
            }
            catch (Exception e)
            {
                Debug.LogWarning("[extNPC] could not write the session file at " +
                                 Path + ": " + e.Message);
                return false;
            }

            // No bookkeeping needed for our own write: the source check in
            // TryPoll drops it, and it drops a previous process's payload too,
            // which a sequence number cannot because it resets on restart.
            return true;
        }

        /// <summary>
        /// What the OTHER side is looking at, or false when there is no news.
        ///
        /// NEVER THROWS. This runs from an editor update on a timer, and the
        /// file is being replaced by another process, so a missing file, a
        /// half-written one, one being swapped underneath us and a payload from
        /// a future schema all have to mean the same thing: no news. On Windows
        /// a replace can even make the reader's own Open fail transiently,
        /// which is measured and expected and is harmless to a poller.
        /// </summary>
        public bool TryPoll(out State state)
        {
            state = default(State);

            JObject root;
            try
            {
                if (!File.Exists(Path)) return false;
                root = JObject.Parse(File.ReadAllText(Path));
            }
            catch
            {
                return false;
            }

            int? schema = (int?)root["schema"];
            if (!schema.HasValue || schema.Value != Schema) return false;

            string source = (string)root["source"] ?? "";
            if (source == _source) return false;

            long seq = (long?)root["seq"] ?? 0L;
            if (source == _lastSeenSource && seq == _lastSeenSeq) return false;

            string world = (string)root["world"] ?? "";
            // A dashboard on a different world must not drag the viewer to a
            // villager who does not exist in the one on screen. Empty means
            // "not stated", which is allowed so a caller can opt out.
            if (!string.IsNullOrEmpty(_world) && !string.IsNullOrEmpty(world)
                && world != _world)
                return false;

            _lastSeenSource = source;
            _lastSeenSeq = seq;
            state = new State
            {
                Selected = root["selected"] != null && root["selected"].Type != JTokenType.Null
                    ? (string)root["selected"] : null,
                Year = root["year"] != null && root["year"].Type != JTokenType.Null
                    ? (int?)(int)root["year"] : null,
                World = world,
                Source = source,
                Seq = seq,
            };
            return true;
        }

        /// <summary>Formats a year the way the Python side parses it. Invariant
        /// culture on purpose: this machine writes commas for decimal points in
        /// its own locale, and the reader on the other end does not.</summary>
        public static string FormatYear(int year)
        {
            return year.ToString(CultureInfo.InvariantCulture);
        }
    }
}
